nixos/lib/test-driver/test_driver/polling_condition.py

```diff
@@ -0,0 +1,90 @@
+from typing import Callable, Optional, Any, List, Dict
+from functools import wraps
+
+import time
+
+from .logger import rootlog
+
+
+class PollingConditionFailed(Exception):
+    pass
+
+
+def coopmulti(fun: Callable, *, machine: Any = None) -> Callable:
+    assert not (fun is None and machine is None)
+
+    def inner(fun_: Callable) -> Any:
+        @wraps(fun_)
+        def wrapper(*args: List[Any], **kwargs: Dict[str, Any]) -> Any:
+            this_machine = args[0] if machine is None else machine
+
+            if this_machine.fail_early():  # type: ignore
+                raise PollingConditionFailed("Action interrupted early...")
+
+            return fun_(*args, **kwargs)
+
+        return wrapper
+
+    if fun is None:
```

Comment:
i am not sure how `fun` can be `None` if its type annotation says it's `Callable` - can this check be simply removed?

Reply:
Yeah, probably could, the intended usage was to allow `@coopmulti(machine=...)` for non-bound methods, not sure if that's a use-case we should support. If we do, it should be `fun: Optional[Callable]`. If we don't, the whole `machine` argument can be removed.

Reply:
then let's keep it as simple as possible by providing what we need to support right now in order to provide the functionality that this PR tries to introduce.

Reply:
Actually it can be done a lot nicer _without_ `coopmulti`, logic is a lot simpler now and things are closer to where'd you expect them now